tests/ranged_balance.cpp

```diff
@@ -0,0 +1,275 @@
+    }
+    shooter.wield( gun );
+}
+
+static void equip_shooter( npc &shooter, std::vector<std::string> apparel )
+{
+    tripoint shooter_pos( 60, 60, 0 );
+    shooter.setpos( shooter_pos );
+    shooter.worn.clear();
+    shooter.inv.clear();
+    for( const std::string article : apparel ) {
+        shooter.wear_item( item( article ) );
+    }
+}
+
+std::array<double, 5> accuracy_levels = {{ accuracy_grazing, accuracy_standard, accuracy_goodhit, accuracy_critical, accuracy_headshot }};
+
+static std::array<statistics, 5> firing_test( dispersion_sources dispersion, int range,
+        std::array<double, 5> thresholds )
+{
+    std::array<statistics, 5> firing_stats;
+    bool threshold_within_confidence_interval = false;
+    do {
+        projectile_attack_aim aim = projectile_attack_roll( dispersion, range, 0.5 );
+        threshold_within_confidence_interval = false;
+        for( int i = 0; i < accuracy_levels.size(); ++i ) {
+            firing_stats[i].add( aim.missed_by < accuracy_levels[i] );
+            if( thresholds[i] == -1 ) {
+                continue;
+            }
+            if( firing_stats[i].n() < 100 ) {
+                threshold_within_confidence_interval = true;
+            }
+            double error = firing_stats[i].adj_wald_error();
+            double avg = firing_stats[i].avg();
+            double threshold = thresholds[i];
+            if( avg + error > threshold && avg - error < threshold ) {
+                threshold_within_confidence_interval = true;
+            }
+        }
+    } while( threshold_within_confidence_interval && firing_stats[0].n() < 100000 );
+    return firing_stats;
+}
+
+static dispersion_sources get_dispersion( npc &shooter, int aim_time )
+{
+    item &gun = shooter.weapon;
+    dispersion_sources dispersion = shooter.get_weapon_dispersion( gun );
+
+    // The 10 is an arbitrary amount under which NPCs refuse to spend moves on aiming.
+    shooter.moves = 10 + aim_time;
+    shooter.recoil = MAX_RECOIL;
+    // Aim as well as possible within the provided time.
+    shooter.aim();
+    if( aim_time > 0 ) {
+        REQUIRE( shooter.recoil < MAX_RECOIL );
+    }
+    dispersion.add_range( shooter.recoil );
+
+    return dispersion;
+}
+
+static void test_shooting_scenario( npc &shooter, int min_quickdraw_range,
+                                    int min_good_range, int max_good_range )
+{
+    {
+        dispersion_sources dispersion = get_dispersion( shooter, 0 );
+        std::array<statistics, 5> minimum_stats = firing_test( dispersion, min_quickdraw_range, {{ 0.2, 0.1, -1, -1, -1 }} );
+        INFO( dispersion );
+        INFO( "Range: " << min_quickdraw_range );
+        INFO( "Max aim speed: " << shooter.aim_per_move( shooter.weapon, MAX_RECOIL ) );
+        INFO( "Min aim speed: " << shooter.aim_per_move( shooter.weapon, shooter.recoil ) );
+        CHECK( minimum_stats[0].avg() < 0.2 );
+        CHECK( minimum_stats[1].avg() < 0.1 );
```

Comment:
Those unnamed arrays could use an enum to address them. `minimum_stats[GRAZE]` is so much clearer than `minimum_stats[0]`.